python/lsst/resources/http.py

```diff
@@ -252,26 +247,79 @@ def _get_temp_dir() -> Tuple[str, int]:
+        # token may be the token value itself or a path to a file containing
+        # the token value. The file must be protected so that only its owner
+        # can access it
+        self._token = self._path = None
+        self._mtime: float = -1.0
+        if not token:
+            return
+        self._token = token
+        if os.path.isfile(token):
+            self._path = os.path.abspath(token)
+            if not _is_protected(self._path):
+                raise PermissionError(
+                    f"Bearer token file at {self._path} must be protected for access only by its owner"
+                )
+            self._refresh()
+
+    def _refresh(self) -> None:
+        # Read the token file (if any) if its modification time is more recent
+        # than the the last time we read it
+        if not self._path:
+            return
+
+        if (mtime := os.stat(self._path).st_mtime) > self._mtime:
+            log.debug("Reading bearer token file at %s", self._path)
+            self._mtime = mtime
+            with open(self._path) as f:
+                self._token = f.read().rstrip("\n")
+
+    def __call__(self, req: requests.PreparedRequest) -> requests.PreparedRequest:
+        if self._token:
+            self._refresh()
+            req.headers["Authorization"] = f"Bearer {self._token}"
+        return req
+
+
 class HttpResourcePath(ResourcePath):
     """General HTTP(S) resource."""
 
-    _session = requests.Session()
-    _sessionInitialized = False
     _is_webdav: Optional[bool] = None
 
+    # Use a session exclusively for PUT requests and another session for
+    # all other requests. PUT requests may be redirected and in that case
+    # the server may close the persisted connection. If that is the case
+    # only the connection persisted for PUT requests will be closed and
+    # the other persisted connection will be kept alive and reused for
+    # other requests.
+    _session: Optional[requests.Session] = None
+    _upload_session: Optional[requests.Session] = None
+
     @property
     def session(self) -> requests.Session:
         """Client object to address remote resource."""
         cls = type(self)
-        if cls._sessionInitialized:
-            if isTokenAuth():
-                refreshToken(cls._session)
+        if cls._session:
             return cls._session
 
-        s = getHttpSession()
-        cls._session = s
-        cls._sessionInitialized = True
-        return s
+        cls._session = _get_http_session(self)
+        return cls._session
+
+    @property
+    def upload_session(self) -> requests.Session:
+        """Client object to address remote resource for PUT requests."""
+        cls = type(self)
+        if cls._upload_session:
+            return cls._upload_session
+
+        log.debug("Creating new HTTP session for PUT requests: %s", self.geturl())
+        cls._upload_session = _get_http_session(self)
+        return cls._upload_session
```

Comment:
Similar comment above regarding using a class attribute as an instance property when the root URI can change.

Reply:
You are right. I was not satisfied with that solution but thought it was the initial design and did not attempt to improve it.

I introduced a new class `SessionStore` which caches and makes reusable client sessions, one per endpoint. This gives more flexibility but there are still things to improve. For instance, the bearer token for authenticating to a remote endpoint is passed in as an environment variable `LSST_HTTP_AUTH_BEARER_TOKEN`. This means that one can only have a single token for all the remotes, which is not what happens in the practice. A similar issue happens with the client certificate to present to the server (when using X509 certificates as the client authentication method): currently one can only provide one client certificate and it is used for all the remote endpoints.

I think this should be improved in a separate ticket.